Adding erxupdate script to nprobe repository

swilkey opened the pull request:
Hi Luca and team,

This is a pull request to add the ntopnghosts.sh script (currently v1.5) to the nprobe repository.  I've stored this in a directory called erxupdate.

I hope this is useful and I've made this request in the right way - I've never done github before.

Regards,
Stephen

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+	if [[ "$NAMESUBSTITUTESIZE" != "0" ]]; then	
+		if [[ "$FILESIZE" != "$NAMESUBSTITUTESIZE" ]]; then
+			echo "Downloading updated name substitute file"
+			curl -s $NAMESUBSTITUTEURL -o "/tmp/hostnamesubstitute_new"
```
This can open out to a series of security issues. I would at least force curl to use TLS. An attacker can trick you into storing data under tmp. Then, this data can be used later to inject shell code as it is not checked against any special character.

swilkey replied:
updated curl commands to use --ssl
I don't want to use --ssl-reqd because if someone has a non-https server that they need to use then they would be unable to use the script.  I have added a note in the instructions recommending that https servers be used for hosting the script and hostnamesubstitute file

swilkey replied:
I'll create a folder under /run to replace my use of /tmp.  This will ensure it is still on a tmpfs (RAM not long term storage to prevent wear of long term storage) and it will hopefully meet your requirement of not using the /tmp folder to reduce risk of attackers tricking us to using injected data in the tmp folder.

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+			echo "Downloading updated name substitute file"
+			curl -s $NAMESUBSTITUTEURL -o "/tmp/hostnamesubstitute_new"
+			FILESIZE=`wc -c /tmp/hostnamesubstitute_new|cut -d ' ' -f1`
+			if [[ "$FILESIZE" == "$NAMESUBSTITUTESIZE" ]]; then
```
Use an hash function rather than the size to determine changes in a more reliable way.

swilkey replied:
good idea but I don't want to bog down a router with cryptographic functions computing hashes of files, so instead I'll switch to using cmp to compare the file contents for any change directly.

swilkey replied:
Additional note to the above response.  My comments were in relation to all the places in the script where I was using the filesize to determine if the file was the same or not.  However, they did not relate specifically to the instance above.  In the instance above we are checking the filesize on the remote server and comparing it with the local filesize so as to decide whether to download the remote file or not.  To my knowledge there is no way to compute a hash of a remote file without downloading it, and it was a design parameter not to download anything unnecessarily and therefore not to download it just to determine not to download it.  As a result, for determining whether to download the file or not we have to stick with using the filesize as that information IS available without downloading the file.

swilkey replied:
The other time we use the filesize for comparison is immediately after downloading the file we confirm that we have downloaded a file of the size that we were anticipating.  This is to protect against the file download being interrupted and receiving an incomplete file and then accidentally using that incomplete file. Again, this is not suitable for replacing with a hash function as to do so we would have to know the original hash of the file on the remote server - which is not possible to my knowledge.

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+            SUB=`grep -i "^$HOSTNAME $MAC" /tmp/hostnamesubstitute | cut -d ' ' -f 3-100`
+            if [[ "$SUB" != "" ]]; then
+                echo "Replacing $HOSTNAME with $SUB"
+                NTOPNGHOSTNAME="$SUB"
```
SUB can contain injection chars if it comes from an untrusted source, causing arbitrary shell code execution.

swilkey replied:
SUB doesn't come from an untrusted source.  SUB is derived from a line in /tmp/hostnamesubstitute.  Hostnamesubstitute is either sourced from /etc/nprobe/ which is safe within the router (unless someone hacks the router - in which case the system is fully compromised already) or from a webserver controlled by the person operating the router - they define the URL that they trust to download this from and they define the contents of the file on the webserver.  If they don't secure the webserver and the webserver gets hacked then there isn't anything we can do to protect them as they already have a big problem.

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+            fi
+        fi
+        # submit the host to NTOPNG
+        RESULT=`curl -s -u $NTOPNGUSER:$NTOPNGPASS -H "Content-Type: application/json" -d '{"host": "'$IP'", "custom_name": "'$NTOPNGHOSTNAME'"}' "https://ntopng.iam-afghanistan.org/lua/rest/v1/set/host/alias.lua"`
```
I believe this url should be parametric.

swilkey replied:
Why?  The server name is static, the rest parameters are static- what advantage would there be changing:
https://ntopng.iam-afghanistan.org/lua/rest/v1/set/host/alias.lua
to
$CONNECTION=https://
$HOSTNAME=ntopng.iam-afghanistan.org
$PATH=/lua/rest/v1/set/host/alias.lua
$CONNECTION$HOSTNAME$PATH
?
Happy to make a change if it makes sense but that just looks like unnecessary code complication.

simonemainardi replied:
I mean `https://ntopng.iam-afghanistan.org` should be a parameter. Not sure all script users want to post data to `https://ntopng.iam-afghanistan.org`

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+			# changes IP address as we'll never retry again.  The only thing that could cause a failure is if we have no internet connection when we run this 
+			# or if the ntopng server is down.  Note that following a reboot of the router we WILL try again to register all known hosts from the hosts file 
+			# to the ntopng server
+            RESULT=`curl -s -u $NTOPNGUSER:$NTOPNGPASS -H "Content-Type: application/json" -d '{"host": "'$IP'", "custom_name": "'$NTOPNGHOSTNAME'"}' "https://ntopng.iam-afghanistan.org/lua/rest/v1/set/host/alias.lua"`
```
I believe this url should be parametric.

swilkey replied:
See comment above.

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+	if [[ "$UPDATESIZE" != "0" ]]; then
+		if [[ "$FILESIZE" != "$UPDATESIZE" ]]; then
+			echo "Downloading updated script"
+			curl -s $UPDATEURL -o "/tmp/ntopnghosts.sh"
```
This can open up to a series of security issues as discussed above.

swilkey replied:
switched to --ssl to require ssl upgrade

simonemainardi commented on erxupdate/ntopnghosts.sh:
```diff
@@ -0,0 +1,429 @@
+			FILESIZE=`wc -c /tmp/ntopnghosts.sh|cut -d ' ' -f1`
+			if [[ "$FILESIZE" == "$UPDATESIZE" ]]; then
+				echo 'Script is up-to-date'
+				cp /tmp/ntopnghosts.sh /etc/nprobe/ntopnghosts.sh
```
This can be very dangerous as:
1. You are trusting an executable downloaded from the internet and copying it under /etc. Should really be sure the executable come from a trusted source.
1. You are trusting an executable from /tmp. Note that /tmp is readable and writable by anyone. An attacker could write `/tmp/ntopnghosts.sh` and trick you into readying it and copying to /etc. I would create a script-specific folder for the script to be used for temporary data and only give read-write privileges to the user running this script.

swilkey replied:
Script will be updated to use /run instead of /tmp.
The executable comes from the source url defined by the person who installs the script.  If they don't secure their webserver then they have a bigger problem and we cannot protect against that.  If people don't want to use centrally managed scripts then they can choose not to host the script on a webserver for centrally updating from.
The user running this script needs to be root - we run it using sudo.  This is to deal with the fact that we don't want to be changing permissions on the router and breaking any of the router's functions in doing so.

simonemainardi reviewed (changes requested):
Thanks for the contribution. I believe this is a nice integration that shows how to fetch data from routers and push it into ntopng. Please, review according to my comments.

swilkey commented:
Hi Simone, thanks for your detailed comments.  I've responded to them and made a number of updates to the script to address your concerns.  The one remaining concern that I don't feel comfortable I've addressed is the comment about parametric urls.  If you can respond to my comment about that maybe I'll understand what you were asking for.  

Once that is also addressed I'm not sure what to do next.  Do I need to click resolve conversation next to each of your comments?  Do I need to close the pull request, make my changes in my branch and then submit a new pull request?  This is the first time I've used github and done a pull request so I'm unsure how to proceed.

swilkey commented:
OOPS!  You are right.  thanks for picking that up!

Once I've corrected that, how do I do the next step?  Cancel the pull 
request, update the script and issue a new pull request?

------ Original Message ------
From: "Simone Mainardi" ***@***.***>
To: "ntop/nProbe" ***@***.***>
Cc: "swilkey" ***@***.***>; "Author" 
***@***.***>
Sent: 19/04/2021 8:03:42 PM
Subject: Re: [ntop/nProbe] Adding erxupdate script to nprobe repository 
(#473)


@simonemainardi commented on this pull request.

--------------------------------------------------------------------------------
In erxupdate/ntopnghosts.sh 
<https://github.com/ntop/nProbe/pull/473#discussion_r615954091>:

 > +	# reporting to telegram if you have enabled that functionality.
+	MAC=`echo "$line" | cut -d ' ' -f 2`
+    echo "Checking IP=$IP HOSTNAME=$HOSTNAME MAC=$MAC"
+	# echo "line=$line"
+	if [[ "$HOSTNAME $MAC" != "$LASTHOSTNAME $LASTMAC" ]]; then
+    	NTOPNGHOSTNAME="$HOSTNAME($MAC)"
+        # check if we need to replace the hostname with something more meaningful that we have on file
+		if [[ -f "/tmp/hostnamesubstitute" ]]; then
+            SUB=`grep -i "^$HOSTNAME $MAC" /tmp/hostnamesubstitute | cut -d ' ' -f 3-100`
+            if [[ "$SUB" != "" ]]; then
+                echo "Replacing $HOSTNAME with $SUB"
+                NTOPNGHOSTNAME="$SUB"
+            fi
+        fi
+        # submit the host to NTOPNG
+        RESULT=`curl -s -u $NTOPNGUSER:$NTOPNGPASS -H "Content-Type: application/json" -d '{"host": "'$IP'", "custom_name": "'$NTOPNGHOSTNAME'"}' "https://ntopng.iam-afghanistan.org/lua/rest/v1/set/host/alias.lua"`

I mean https://ntopng.iam-afghanistan.org should be a parameter. Not 
sure all script users want to post data to 
https://ntopng.iam-afghanistan.org

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub 
<https://github.com/ntop/nProbe/pull/473#discussion_r615954091>, or 
unsubscribe 
<https://github.com/notifications/unsubscribe-auth/AB5UXTBOYMXFEVYILXGGOW3TJRENNANCNFSM43EONOFA>.

simonemainardi commented:
Good to go. Please, send the PR under the ntopng repository, under contrib/erxupdate and it will be integrated into the dev branch.
